src/workflow/CommandExecutor.py

```diff
@@ -0,0 +1,193 @@
+            threads.append(thread)
+
+        # Wait for all threads to complete
+        for thread in threads:
+            thread.join()
+
+        # Calculate and log the total execution time
+        end_time = time.time()
+        self.logger.log(
+            f"Total time to run {len(commands)} commands: {end_time - start_time:.2f} seconds"
+        )
+
+    def run_command(self, command: list[str], write_log: bool = True) -> None:
+        """
+        Executes a specified shell command and logs its execution details.
+
+        Args:
+            command (list[str]): The shell command to execute, provided as a list of strings.
+            write_log (bool): If True, logs the command's output and errors.
+
+        Raises:
+            Exception: If the command execution results in any errors.
+        """
+        # Ensure all command parts are strings
+        command = [str(c) for c in command]
+        
+        # Log the execution start
+        self.logger.log(f"Running command:\n"+' '.join(command)+"\nWaiting for command to finish...")
+        
+        start_time = time.time()
+        
+        # Execute the command
+        process = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        child_pid = process.pid
+        
+        # Record the PID for potential management
+        pid_file_path = self.pid_dir / str(child_pid)
+        pid_file_path.touch()
+        
+        # Wait for command completion and capture output
+        stdout, stderr = process.communicate()
+        
+        # Cleanup PID file
+        pid_file_path.unlink()
+        
+        end_time = time.time()
+        execution_time = end_time - start_time
+        
+        # Format the logging prefix
+        self.logger.log(f"Total time to run command: {execution_time:.2f} seconds")
+        
+        # Log stdout if present
+        if stdout and write_log:
+            self.logger.log(f"Console log:\n\n{stdout.decode()}")
+        
+        # Log stderr and raise an exception if errors occurred
+        if stderr or process.returncode != 0:
+            error_message = stderr.decode().strip()
+            self.logger.log(f"ERRORS OCCURRED:\n{error_message}")
+            raise Exception(f"Errors occurred while running command: {' '.join(command)}\n{error_message}")
+
+    def run_topp(self, tool: str, input_output: dict, write_log: bool = True) -> None:
+        """
+        Constructs and executes commands for the specified tool OpenMS TOPP tool based on the given
+        input and output configurations. Ensures that all input/output file lists
+        are of the same length, or single strings, to maintain consistency in command
+        execution. Supports executing commands either as single or multiple processes
+        based on the input size.
+
+        Args:
+            tool (str): The executable name or path of the tool.
+            input_output (dict): A dictionary specifying the input and output
```

Comment:
Suggestion (because it was a bit confusing to me) to change:
A dictionary specifying the input/output parameter names (as key) and their corresponding file paths (as value).